crates/ra_hir_ty/src/expr.rs

```diff
@@ -310,3 +310,25 @@ pub fn record_pattern_missing_fields(
     }
     Some((variant_def, missed_fields, exhaustive))
 }
+
+pub fn unsafe_expressions(
+    db: &dyn HirDatabase,
+    infer: &InferenceResult,
+    def: DefWithBodyId,
+) -> Vec<ExprId> {
+    let mut unsafe_expr_ids = vec![];
+    let body = db.body(def);
+    for (id, expr) in body.exprs.iter() {
+        if let Expr::Call { callee, .. } = expr {
+            if infer
+                .method_resolution(*callee)
+                .map(|func| db.function_data(func).is_unsafe)
+                .unwrap_or(false)
+            {
+                unsafe_expr_ids.push(id);
+            }
+        }
+    }
```

Comment:
This is missing raw pointer dereferences. Also taking references to packed structs should use unsafe in most cases. (currently always allowed without unsafe because of a not yet patched soundness hole)

Reply:
Just accounted for derefencing raw pointers. Regarding references to packed structs, is there a way to figure out which of the "most cases" are invalid?

Reply:
It's references to *fields of* packed structs, right? This seems like it might require some better tracking of place expressions, plus we'd need to keep track of struct representation in the first place.

Reply:
> It's references to fields of packed structs, right?

Yes

Reply:
Should I go ahead and track `Attrs` along with `StructData` in this work, or would it be easier to review as a follow-up addition?

Reply:
I'm always for keeping PRs small ;)

Reply:
Created https://github.com/rust-analyzer/rust-analyzer/issues/4600 to track that separately. Cross referenced back here.